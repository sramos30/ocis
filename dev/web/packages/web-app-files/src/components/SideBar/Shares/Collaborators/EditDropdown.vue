<template>
  <span class="oc-flex oc-flex-middle">
    <oc-button :id="editShareBtnId" class="collaborator-edit-dropdown-options-btn" appearance="raw">
      <oc-icon name="more-2" />
    </oc-button>
    <oc-drop
      ref="expirationDateDrop"
      :toggle="'#' + editShareBtnId"
      mode="click"
      padding-size="remove"
    >
      <oc-list class="collaborator-edit-dropdown-options-list" :aria-label="shareEditOptions">
        <li v-if="isExpirationSupported" class="oc-px-s oc-py-xs">
          <oc-datepicker
            v-model="enteredExpirationDate"
            :min-date="minExpirationDate"
            :max-date="maxExpirationDate"
            :locale="$language.current"
            :is-required="isExpirationDateEnforced"
            class="files-recipient-expiration-datepicker"
            data-testid="recipient-datepicker"
          >
            <template #default="{ togglePopover }">
              <oc-button
                class="files-collaborators-expiration-button"
                data-testid="recipient-datepicker-btn"
                appearance="raw"
                @click="togglePopover"
              >
                <span v-if="isExpirationDateSet" v-text="$gettext('Edit expiration date')" />
                <span v-else v-text="$gettext('Set expiration date')" />
              </oc-button>
            </template>
          </oc-datepicker>
        </li>
        <li v-for="(option, i) in options" :key="i" class="oc-px-s oc-py-xs">
          <oc-button
            appearance="raw"
            :class="option.class"
            v-bind="option.additionalAttributes || {}"
            @click="option.method()"
          >
            <span v-text="option.title" />
          </oc-button>
        </li>
      </oc-list>
    </oc-drop>
  </span>
</template>

<script>
import { mapGetters } from 'vuex'
import { DateTime } from 'luxon'
import Mixins from '../../../../mixins'

export default {
  name: 'EditDropdown',
  mixins: [Mixins],
  props: {
    expirationDate: {
      type: Date,
      required: false,
      default: undefined
    },
    shareCategory: {
      type: String,
      required: false,
      default: null,
      validator: function (value) {
        return ['user', 'group'].includes(value) || !value
      }
    }
  },
  data: function () {
    return {
      enteredExpirationDate: null
    }
  },
  computed: {
    ...mapGetters(['capabilities']),

    options() {
      const result = []
      if (this.isRemoveExpirationPossible) {
        result.push({
          title: this.$gettext('Remove expiration date'),
          method: this.removeExpirationDate,
          class: 'remove-expiration-date',
          additionalAttributes: {
            'data-testid': 'collaborator-remove-expiration-btn'
          }
        })
      }
      return [
        ...result,
        {
          title: this.$gettext('Remove share'),
          method: this.removeShare,
          class: 'remove-share',
          additionalAttributes: {
            variation: 'danger',
            'data-testid': 'collaborator-remove-share-btn'
          }
        }
      ]
    },

    editShareBtnId() {
      return 'files-collaborators-edit-button-' + this._uid
    },
    shareEditOptions() {
      return this.$gettext('Context menu of the share')
    },

    editingUser() {
      return this.shareCategory === 'user'
    },

    editingGroup() {
      return this.shareCategory === 'group'
    },

    isExpirationSupported() {
      return (
        (this.editingUser && this.userExpirationDate) ||
        (this.editingGroup && this.groupExpirationDate)
      )
    },

    isExpirationDateSet() {
      return !!this.expirationDate
    },

    isRemoveExpirationPossible() {
      return (
        this.isExpirationSupported && this.isExpirationDateSet && !this.isExpirationDateEnforced
      )
    },

    isDefaultExpirationEnabled() {
      if (this.editingUser) {
        return this.userExpirationDate.enabled
      }

      if (this.editingGroup) {
        return this.groupExpirationDate.enabled
      }

      return this.userExpirationDate.enabled || this.groupExpirationDate.enabled
    },

    userExpirationDate() {
      return this.capabilities.files_sharing.user.expire_date
    },

    groupExpirationDate() {
      return this.capabilities.files_sharing.group?.expire_date
    },

    defaultExpirationDate() {
      if (!this.isDefaultExpirationEnabled) {
        return null
      }

      const userMaxExpirationDays = parseInt(this.userExpirationDate.days)
      const groupMaxExpirationDays = parseInt(this.groupExpirationDate.days)
      let days = 0

      if (this.editingUser) {
        days = userMaxExpirationDays
      } else if (this.editingGroup) {
        days = groupMaxExpirationDays
      } else if (userMaxExpirationDays && groupMaxExpirationDays) {
        days = Math.min(userMaxExpirationDays, groupMaxExpirationDays)
      } else {
        days = userMaxExpirationDays || groupMaxExpirationDays
      }

      const date = new Date()
      date.setDate(new Date().getDate() + days)
      return date
    },

    isExpirationDateEnforced() {
      if (this.editingUser) {
        return this.userExpirationDate.enforced
      }

      if (this.editingGroup) {
        return this.groupExpirationDate.enforced
      }

      return this.userExpirationDate.enforced || this.groupExpirationDate.enforced
    },

    maxExpirationDate() {
      if (!this.isExpirationDateEnforced) {
        return null
      }
      return this.defaultExpirationDate
    },

    minExpirationDate() {
      const date = new Date()
      date.setDate(new Date().getDate() + 1)
      return date
    },

    relativeExpirationDate() {
      return DateTime.fromJSDate(this.enteredExpirationDate)
        .setLocale(this.$language.current)
        .endOf('day')
        .toRelative()
    }
  },
  watch: {
    enteredExpirationDate: {
      handler: 'updateExpirationDate'
    }
  },
  methods: {
    updateExpirationDate() {
      this.$emit('expirationDateChanged', {
        expirationDate: DateTime.fromJSDate(this.enteredExpirationDate).endOf('day').toISO()
      })
      this.$refs.expirationDateDrop.hide()
    },
    removeExpirationDate() {
      this.$emit('expirationDateChanged', {
        expirationDate: null
      })
      this.$refs.expirationDateDrop.hide()
    },
    removeShare() {
      this.$emit('removeShare')
    }
  }
}
</script>
