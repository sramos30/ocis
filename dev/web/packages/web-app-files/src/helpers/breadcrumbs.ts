import { bus } from 'web-pkg/src/instance'

export interface BreadcrumbItem {
  text: string
  to?: string
  allowContextActions?: boolean
  onClick?: () => void
}

export const breadcrumbsFromPath = (
  currentPath: string,
  resourcePath: string
): BreadcrumbItem[] => {
  const pathSplit = (p = '') => p.split('/').filter(Boolean)
  const current = pathSplit(currentPath)
  const resource = pathSplit(resourcePath)

  return resource.map(
    (text, i) =>
      ({
        allowContextActions: true,
        text,
        to: '/' + [...current].splice(0, current.length - resource.length + i + 1).join('/')
      } as BreadcrumbItem)
  )
}

export const concatBreadcrumbs = (...items: BreadcrumbItem[]): BreadcrumbItem[] => {
  const last = items.pop()

  return [
    ...items,
    {
      allowContextActions: last.allowContextActions,
      text: last.text,
      onClick: () => bus.publish('app.files.list.load')
    }
  ]
}
