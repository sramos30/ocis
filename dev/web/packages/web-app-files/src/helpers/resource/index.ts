export * from './asset'
export * from './filter'
export * from './privatePreviewBlob'
export * from './publicPreviewUrl'
export * from './resource'
export * from './sameResource'
