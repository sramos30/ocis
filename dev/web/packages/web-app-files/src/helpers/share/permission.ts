// dummy to trick gettext string extraction into recognizing strings
const $gettext = (str) => {
  return str
}

export class SharePermission {
  private readonly _key: string
  private readonly _bit: number
  private readonly _label: string

  constructor(key: string, bit: number, label: string) {
    this._key = key
    this._bit = bit
    this._label = label
  }

  get key(): string {
    return this._key
  }

  get bit(): number {
    return this._bit
  }

  get label(): string {
    return $gettext(this._label)
  }

  public enabled(bitmask: number): boolean {
    return (bitmask & this.bit) > 0
  }
}

export abstract class SharePermissions {
  static readonly read = new SharePermission('read', 1, $gettext('Read'))
  static readonly update = new SharePermission('update', 2, $gettext('Update'))
  static readonly create = new SharePermission('create', 4, $gettext('Create'))
  static readonly delete = new SharePermission('delete', 8, $gettext('Delete'))
  static readonly share = new SharePermission('share', 16, $gettext('Share'))

  static permissionsToBitmask(permissions: SharePermission[]): number {
    return (permissions || []).reduce((b: number, p: SharePermission) => b | p.bit, 0)
  }

  static bitmaskToPermissions(bitmask: number): SharePermission[] {
    return [this.read, this.update, this.create, this.delete, this.share].reduce(
      (a: SharePermission[], p: SharePermission) => {
        if (p.enabled(bitmask)) {
          a.push(p)
        }
        return a
      },
      []
    )
  }
}
