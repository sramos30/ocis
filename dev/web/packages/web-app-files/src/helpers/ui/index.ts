export * as fileList from './filesList'
export * as resourceTable from './resourceTable'
