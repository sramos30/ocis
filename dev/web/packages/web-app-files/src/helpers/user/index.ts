export * from './avatarUrl'
export * from './types'
