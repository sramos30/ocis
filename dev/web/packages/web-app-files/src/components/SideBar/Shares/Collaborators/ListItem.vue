<template>
  <div
    :data-testid="`collaborator-${isUser || isSpace ? 'user' : 'group'}-item-${
      share.collaborator.name
    }`"
    class="files-collaborators-collaborator oc-flex oc-flex-middle oc-py-xs oc-flex-between"
  >
    <div class="oc-width-2-3 oc-flex oc-flex-middle" style="gap: 10px">
      <avatar-image
        v-if="isUser || isSpace"
        :userid="share.collaborator.name"
        :user-name="share.collaborator.displayName"
        :width="48"
        class="files-collaborators-collaborator-indicator"
      />
      <oc-avatar-item
        v-else
        :width="48"
        icon-size="medium"
        :icon="shareTypeKey"
        :name="shareTypeKey"
        class="files-collaborators-collaborator-indicator"
      />
      <div class="oc-text-truncate">
        <p v-oc-tooltip="shareDisplayNameTooltip" class="oc-text-bold oc-text-truncate oc-m-rm">
          <span
            aria-hidden="true"
            class="files-collaborators-collaborator-name"
            v-text="shareDisplayName"
          />
          <span
            v-if="shareAdditionalInfo"
            aria-hidden="true"
            class="files-collaborators-collaborator-additional-info"
            v-text="shareAdditionalInfo"
          />
          <span class="oc-invisible-sr" v-text="screenreaderShareDisplayName" />
        </p>
        <p class="oc-m-rm oc-flex">
          <span
            aria-hidden="true"
            class="files-collaborators-collaborator-share-type"
            v-text="shareTypeText"
          />
          <span v-if="sharedParentRoute" class="oc-resource-indicators oc-text-truncate">
            <span class="oc-mx-s">Â·</span>
            <router-link
              v-oc-tooltip="$gettext('Navigate to parent folder')"
              class="parent-folder oc-text-truncate"
              :to="sharedParentRoute"
            >
              <span class="text" v-text="$gettext('via')" />
              <oc-icon name="folder-2" size="small" fill-type="line" class="oc-px-xs" />
              <span class="text oc-text-truncate" v-text="sharedParentDir" />
            </router-link>
          </span>
          <span class="oc-invisible-sr" v-text="screenreaderShareDetails" />
        </p>
        <p v-if="hasExpirationDate" class="oc-m-rm">
          <span
            v-oc-tooltip="expirationDate"
            aria-hidden="true"
            class="files-collaborators-collaborator-expiration"
            data-testid="recipient-info-expiration-date"
            tabindex="0"
            v-text="shareExpirationText"
          />
          <span class="oc-invisible-sr" v-text="screenreaderShareExpiration" />
        </p>
      </div>
    </div>
    <div
      v-if="canEditOrDelete"
      class="oc-width-1-3 oc-flex oc-flex-nowrap oc-flex-right oc-flex-middle"
    >
      <role-dropdown
        :resource="highlightedFile"
        :share-id="share.id"
        :existing-permissions="share.customPermissions"
        :existing-role="share.role"
        :allow-share-permission="hasResharing || isSpace"
        class="files-collaborators-collaborator-role"
        @optionChange="shareRoleChanged"
      />
      <edit-dropdown
        class="files-collaborators-collaborator-edit"
        data-testid="collaborator-edit"
        :expiration-date="share.expires ? share.expires : null"
        :share-category="shareCategory"
        @expirationDateChanged="shareExpirationChanged"
        @removeShare="removeShare"
      />
    </div>
    <div v-else-if="share.role">
      <span class="oc-mr-xs" v-text="share.role.label" />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapActions, mapState } from 'vuex'
import Mixins from '../../../../mixins'
import { DateTime } from 'luxon'

import EditDropdown from './EditDropdown.vue'
import RoleDropdown from '../RoleDropdown.vue'
import { SharePermissions, ShareTypes } from '../../../../helpers/share'
import { clientService } from 'web-pkg/src/services'
import { useCapabilityFilesSharingResharing } from 'web-pkg/src/composables'

export default {
  name: 'ListItem',
  components: {
    EditDropdown,
    RoleDropdown
  },
  mixins: [Mixins],
  props: {
    share: {
      type: Object,
      required: true
    },
    modifiable: {
      type: Boolean,
      default: false
    },
    sharedParentRoute: {
      type: Object,
      default: null
    }
  },
  setup() {
    return {
      hasResharing: useCapabilityFilesSharingResharing()
    }
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),
    ...mapGetters(['getToken', 'configuration']),
    ...mapState(['user']),

    shareType() {
      return ShareTypes.getByValue(this.share.shareType)
    },

    shareTypeKey() {
      return this.shareType.key
    },

    isGroup() {
      return this.shareType === ShareTypes.group
    },

    isUser() {
      return this.shareType === ShareTypes.user
    },

    isSpace() {
      return this.shareType === ShareTypes.space
    },

    shareTypeText() {
      return this.$gettext(this.shareType.label)
    },

    shareCategory() {
      return ShareTypes.isIndividual(this.shareType) ? 'user' : 'group'
    },

    shareDisplayName() {
      if (this.user.id === this.share.collaborator.name) {
        return this.$gettextInterpolate(this.$gettext('%{collaboratorName} (me)'), {
          collaboratorName: this.share.collaborator.displayName
        })
      }
      return this.share.collaborator.displayName
    },

    shareAdditionalInfo() {
      if (!this.share.collaborator.additionalInfo) {
        return
      }
      return ` (${this.share.collaborator.additionalInfo})`
    },

    shareDisplayNameTooltip() {
      return this.shareDisplayName + (this.shareAdditionalInfo || '')
    },

    screenreaderShareDisplayName() {
      const context = {
        displayName: this.share.collaborator.displayName,
        ...(this.share.collaborator.additionalInfo && {
          additionalInfo: this.share.collaborator.additionalInfo
        })
      }
      let translated = this.$gettext('Share receiver name: %{ displayName }')
      if (this.shareAdditionalInfo) {
        translated = this.$gettext('Share receiver name: %{ displayName } (%{ additionalInfo })')
      }
      return this.$gettextInterpolate(translated, context)
    },

    screenreaderShareDetails() {
      const translated = this.$gettext('Share type: %{ shareType }')
      return this.$gettextInterpolate(translated, { shareType: this.shareTypeText })
    },

    shareExpirationText() {
      const translated = this.$gettext('Expires %{ expiryDateRelative }')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative
      })
    },

    screenreaderShareExpiration() {
      const translated = this.$gettext('Share expires %{ expiryDateRelative } (%{ expiryDate })')
      return this.$gettextInterpolate(translated, {
        expiryDateRelative: this.expirationDateRelative,
        expiryDate: this.expirationDate
      })
    },

    canEditOrDelete() {
      return this.modifiable
    },

    hasExpirationDate() {
      return this.share.expires
    },

    expirationDate() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toLocaleString(DateTime.DATETIME_FULL)
    },

    expirationDateRelative() {
      return DateTime.fromJSDate(this.share.expires)
        .endOf('day')
        .setLocale(this.$language.current)
        .toRelative()
    },

    graphClient() {
      return clientService.graphAuthenticated(this.configuration.server, this.getToken)
    },

    sharedParentDir() {
      return this.sharedParentRoute?.params?.item.split('/').pop()
    }
  },
  methods: {
    ...mapActions(['showMessage']),
    ...mapActions('Files', ['changeShare']),

    removeShare() {
      this.$emit('onDelete', this.share)
    },

    shareRoleChanged({ role, permissions }) {
      const expirationDate = this.share.expires
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply new permissions'),
          status: 'danger'
        })
      }
    },

    shareExpirationChanged({ expirationDate }) {
      const role = this.share.role
      const permissions = this.share.customPermissions
      try {
        this.saveShareChanges({ role, permissions, expirationDate })
      } catch (e) {
        console.error(e)
        this.showMessage({
          title: this.$gettext('Failed to apply expiration date'),
          status: 'danger'
        })
      }
    },

    saveShareChanges({ role, permissions, expirationDate }) {
      const bitmask = role.hasCustomPermissions
        ? SharePermissions.permissionsToBitmask(permissions)
        : SharePermissions.permissionsToBitmask(
            role.permissions(this.hasResharing || this.shareType === ShareTypes.space)
          )
      this.changeShare({
        client: this.$client,
        graphClient: this.graphClient,
        share: this.share,
        permissions: bitmask,
        expirationDate: expirationDate || ''
      })
    }
  }
}
</script>

<style lang="scss" scoped>
.sharee-avatar {
  min-width: 48px;
}
</style>
