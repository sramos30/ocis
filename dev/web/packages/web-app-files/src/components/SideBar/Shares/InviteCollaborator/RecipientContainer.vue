<template>
  <oc-recipient
    :data-testid="`recipient-container-${formattedRecipient.name}`"
    class="files-share-invite-recipient"
    :recipient="formattedRecipient"
  >
    <template #append>
      <oc-button
        class="files-share-invite-recipient-btn-remove"
        appearance="raw"
        :aria-label="btnDeselectRecipientLabel"
        @click.stop="deselect(recipient)"
      >
        <oc-icon name="close" size="small" />
      </oc-button>
    </template>
  </oc-recipient>
</template>

<script>
import { mapGetters } from 'vuex'
import { avatarUrl } from '../../../../helpers/user'
import { ShareTypes } from '../../../../helpers/share'

export default {
  props: {
    recipient: {
      type: Object,
      required: true
    },
    deselect: {
      type: Function,
      required: false,
      default: null
    }
  },

  data() {
    return {
      formattedRecipient: {
        name: this.recipient.label,
        icon: this.getRecipientIcon(),
        hasAvatar: [ShareTypes.user.value, ShareTypes.space.value].includes(
          this.recipient.value.shareType
        ),
        isLoadingAvatar: true
      }
    }
  },

  computed: {
    ...mapGetters(['configuration', 'getToken', 'capabilities']),

    btnDeselectRecipientLabel() {
      const translated = this.$gettext('Deselect %{name}')

      return this.$gettextInterpolate(translated, { name: this.recipient.label })
    }
  },

  async created() {
    if (this.capabilities.files_sharing.user.profile_picture && this.hasAvatar) {
      try {
        this.formattedRecipient.avatar = await avatarUrl({
          clientService: this.$clientService,
          server: this.configuration.server,
          username: this.recipient.value.shareWith,
          token: this.getToken
        })
      } catch (error) {
        console.error(error)
      }
    }

    this.formattedRecipient.isLoadingAvatar = false
  },

  methods: {
    getRecipientIcon() {
      switch (this.recipient.value.shareType) {
        case ShareTypes.group.value:
          return {
            name: 'group',
            label: this.$gettext('Group')
          }

        case ShareTypes.guest.value:
          return {
            name: 'user_remote',
            label: this.$gettext('Guest user')
          }

        case ShareTypes.remote.value:
          return {
            name: 'user_remote',
            label: this.$gettext('Remote user')
          }

        default:
          return {
            name: 'person',
            label: this.$gettext('User')
          }
      }
    }
  }
}
</script>

<style lang="scss" scoped>
.files-share-invite-recipient {
  margin: 4px 2px 0;
  padding: 0 0.25em;
}
</style>
