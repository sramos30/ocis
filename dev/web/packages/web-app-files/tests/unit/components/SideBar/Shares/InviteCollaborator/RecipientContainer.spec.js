import { createLocalVue, mount } from '@vue/test-utils'
import GetTextPlugin from 'vue-gettext'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'

import Users from '@/__fixtures__/users'
import Collaborators from '@/__fixtures__/collaborators'
import SharedFiles from '@/__fixtures__/sharedFiles'

import RecipientContainer from '@files/src/components/SideBar/Shares/InviteCollaborator/RecipientContainer.vue'
import { ShareTypes } from '@files/src/helpers/share/type'
import { buildSharedResource } from '@files/src/helpers/resources'

jest.mock('@files/src/helpers/user/avatarUrl', () => ({
  avatarUrl: jest.fn().mockReturnValue('avatarUrl')
}))

const user = Users.admin
const collaborators = Collaborators

const localVue = createLocalVue()
localVue.use(DesignSystem)
localVue.use(Vuex)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

describe('InviteCollaborator RecipientContainer', () => {
  describe('renders a recipient with a deselect button', () => {
    it.each(ShareTypes.authenticated)('different recipients for different shareTypes', (type) => {
      const { collaborator } = collaborators.filter((sharee) => sharee.shareType === type.value)[0]
      const recipient = {
        label: collaborator.displayName,
        value: {
          shareType: type.key,
          shareWith: collaborator.name,
          shareWithAdditionalInfo: collaborator.additionalInfo
        }
      }
      const wrapper = getMountedWrapper(recipient)
      expect(wrapper).toMatchSnapshot()
    })
  })
  it('displays an avatar image if capability is present', () => {
    const { collaborator } = collaborators[0]
    const recipient = {
      label: collaborator.displayName,
      value: {
        shareType: ShareTypes.user.value,
        shareWith: collaborator.name,
        shareWithAdditionalInfo: collaborator.additionalInfo
      }
    }
    const wrapper = getMountedWrapper(recipient, true)
    expect(wrapper).toMatchSnapshot()
  })
  it('emits an event if deselect button is clicked', async () => {
    const { collaborator } = collaborators[0]
    const recipient = {
      label: collaborator.displayName,
      value: {
        shareType: ShareTypes.user.value,
        shareWith: collaborator.name,
        shareWithAdditionalInfo: collaborator.additionalInfo
      }
    }
    const wrapper = getMountedWrapper(recipient, true)
    const spyOnDeselect = jest.spyOn(wrapper.vm, 'deselect').mockImplementation()
    const button = wrapper.find('.files-share-invite-recipient-btn-remove')
    await button.trigger('click')
    expect(spyOnDeselect).toHaveBeenCalledTimes(1)
  })
})

const storeOptions = (avatarsEnabled) => {
  return {
    state: {
      user
    },
    modules: {
      Files: {
        namespaced: true,
        getters: {
          highlightedFile: () => buildSharedResource(SharedFiles.json().ocs.data[0], true)
        }
      }
    },
    getters: {
      user: () => user,
      getToken: () => '',
      capabilities: () => {
        return {
          files_sharing: {
            user: {
              profile_picture: avatarsEnabled,
              expire_date: {
                enabled: true,
                days: 10
              }
            },
            group: {
              expire_date: {
                enabled: true,
                days: 10
              }
            }
          }
        }
      }
    }
  }
}

function getMountedWrapper(recipient, avatarsEnabled = false) {
  return mount(RecipientContainer, {
    localVue,
    propsData: {
      recipient,
      deselect: jest.fn()
    },
    store: createStore(avatarsEnabled),
    stubs: {
      OcIcon: true
    },
    computed: {
      configuration: () => ({})
    }
  })
}

function createStore(avatarsEnabled) {
  return new Vuex.Store(storeOptions(avatarsEnabled))
}
