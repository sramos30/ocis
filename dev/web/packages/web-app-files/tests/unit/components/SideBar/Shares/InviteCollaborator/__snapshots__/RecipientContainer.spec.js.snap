// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InviteCollaborator RecipientContainer displays an avatar image if capability is present 1`] = `
<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><span width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Albert Einstein" class="vue-avatar--wrapper oc-avatar oc-recipient-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: rgb(194, 28, 83); font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" data-testid="recipient-avatar"><span class="avatarInitials">AE</span></span>
<p data-testid="recipient-name" class="oc-recipient-name">Albert Einstein</p> <button aria-label="Deselect Albert Einstein" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw">
  <ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub>
</button></span>
`;

exports[`InviteCollaborator RecipientContainer renders a recipient with a deselect button different recipients for different shareTypes 1`] = `<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><ocicon-stub name="person" filltype="fill" accessiblelabel="User" type="span" size="small" variation="passive" color="" data-testid="recipient-icon" class="oc-recipient-icon"></ocicon-stub> <p data-testid="recipient-name" class="oc-recipient-name">Albert Einstein</p> <button aria-label="Deselect Albert Einstein" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub></button></span>`;

exports[`InviteCollaborator RecipientContainer renders a recipient with a deselect button different recipients for different shareTypes 2`] = `<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-physics-lovers"><ocicon-stub name="person" filltype="fill" accessiblelabel="User" type="span" size="small" variation="passive" color="" data-testid="recipient-icon" class="oc-recipient-icon"></ocicon-stub> <p data-testid="recipient-name" class="oc-recipient-name">physics-lovers</p> <button aria-label="Deselect physics-lovers" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub></button></span>`;

exports[`InviteCollaborator RecipientContainer renders a recipient with a deselect button different recipients for different shareTypes 3`] = `<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-guest-user"><ocicon-stub name="person" filltype="fill" accessiblelabel="User" type="span" size="small" variation="passive" color="" data-testid="recipient-icon" class="oc-recipient-icon"></ocicon-stub> <p data-testid="recipient-name" class="oc-recipient-name">guest-user</p> <button aria-label="Deselect guest-user" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub></button></span>`;

exports[`InviteCollaborator RecipientContainer renders a recipient with a deselect button different recipients for different shareTypes 4`] = `<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-remote-user"><ocicon-stub name="person" filltype="fill" accessiblelabel="User" type="span" size="small" variation="passive" color="" data-testid="recipient-icon" class="oc-recipient-icon"></ocicon-stub> <p data-testid="recipient-name" class="oc-recipient-name">remote-user</p> <button aria-label="Deselect remote-user" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub></button></span>`;

exports[`InviteCollaborator RecipientContainer renders a recipient with a deselect button different recipients for different shareTypes 5`] = `<span class="oc-recipient files-share-invite-recipient" data-testid="recipient-container-Albert Einstein"><ocicon-stub name="person" filltype="fill" accessiblelabel="User" type="span" size="small" variation="passive" color="" data-testid="recipient-icon" class="oc-recipient-icon"></ocicon-stub> <p data-testid="recipient-name" class="oc-recipient-name">Albert Einstein</p> <button aria-label="Deselect Albert Einstein" type="button" class="files-share-invite-recipient-btn-remove oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-passive oc-button-passive-raw"><ocicon-stub name="close" filltype="fill" accessiblelabel="" type="span" size="small" variation="passive" color=""></ocicon-stub></button></span>`;
